src/librustc_mir/transform/deaggregator.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use rustc::ty::TyCtxt;
+use rustc::mir::repr::*;
+use rustc::mir::transform::{MirPass, MirSource, Pass};
+use rustc_data_structures::indexed_vec::Idx;
+use rustc::ty::{Ty, VariantKind, AdtDef};
+
+pub struct Deaggregator;
+
+impl Pass for Deaggregator {}
+
+impl<'tcx> MirPass<'tcx> for Deaggregator {
+    fn run_pass<'a>(&mut self, tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                    source: MirSource, mir: &mut Mir<'tcx>) {
+        let node_id = source.item_id();
+        let node_path = tcx.item_path_str(tcx.map.local_def_id(node_id));
+        debug!("running on: {:?}", node_path);
+        // we only run when mir_opt_level > 1
+        match tcx.sess.opts.debugging_opts.mir_opt_level {
+            Some(0) |
+            Some(1) |
+            None => { return; },
+            _ => {}
+        };
+        if let MirSource::Fn(_) = source {} else { return; }
+
+        for bb in mir.basic_blocks_mut() {
+            while let Some(ai) = get_aggregate_statement(tcx, &bb.statements) {
+                // do the replacement
+                debug!("removing statement {:?}", ai.statement_index);
+                let src_info = bb.statements[ai.statement_index].source_info;
+                bb.statements.remove(ai.statement_index);
+                for (i, (op, ty)) in ai.operands.iter().zip(ai.types.iter()).enumerate() {
+                    let rhs = Rvalue::Use(op.clone());
+                    let lhs_cast = ai.lhs.clone();
+                    // let lhs_cast = if ai.adt_def.variants.len() > 1 {
+                    //     Lvalue::Projection(Box::new(LvalueProjection {
+                    //         base: ai.lhs.clone(),
+                    //         elem: ProjectionElem::Downcast(ai.adt_def, ai.variant),
+                    //     }))
+                    // } else {
+                    //     ai.lhs.clone()
+                    // };
+                    let lhs_proj = Lvalue::Projection(Box::new(LvalueProjection {
+                        base: lhs_cast,
+                        elem: ProjectionElem::Field(Field::new(i), ty),
+                    }));
+                    let new_statement = Statement {
+                        source_info: src_info,
+                        kind: StatementKind::Assign(lhs_proj, rhs),
+                    };
+                    debug!("inserting: {:?} @ {:?}", new_statement, ai.statement_index + i);
+                    bb.statements.insert(ai.statement_index + i, new_statement);
```

Comment:
Similarly, suspiciously `O(n²)`. On top of the `remove` call, you gonna reposition most of the statements vector for every field in every aggregate in the block.

How about doing something like this instead:

```
let len = bb.statements.len()
let old_statements = ::std::mem::replace(&mut bb.statements, Vec::with_capacity(len));
for statement in statements {
    if !is_applicable_for_optimisation(statement) {
         bb.statements.push(statement);
    } else {
         // do stuff and push your assignments to bb.statements
    }
}
```

While this requires `o(2n)` space, it will avoid multiple cases of potential `O(n²)` time complexity.